Improve default handling for "service create"

aaronlehmann opened the pull request:
Currently, the CLI hardcodes several defaults. This is a problem because it's awkward to keep CLI defaults in sync with swarmkit defaults. Also, when a service is being created, it generally takes on several default values of flags that the user did not specify. If these defaults are changed later in swarmkit, they will not apply to those services, because they will continue to use whatever the CLI hardcoded at the time.

This PR makes two changes across two commits.

1) Changes `docker service inspect` to use the new `InsertDefaults` flag so that fields that are omitted in the service definition show default values in `docker service inspect`. This paves the way for `docker service create` to avoid hardcoding its defaults in services without hiding information from `inspect`.

2) Changes `docker service create` to omit fields that are not specified by the user, when possible. This will allow defaults to be applied inside the manager. To provide visibility into defaults in the CLI help, the CLI uses swarmkit's `api/defaults` package to annotate `docker service create --help` output. These defaults are left out of `docker service update --help`, because defaults are misleading in this context - if a flag is not specified, it has no default action.

Fixes #24959
Fixes #32094

ping @aluzzardi @dongluochen @dnephin @thaJeztah

Commit:
Change "service inspect" to show defaults in place of empty fields

This adds a new parameter insertDefaults to /services/{id}. When this is
set, an empty field (such as UpdateConfig) will be populated with
default values in the API response. Make "service inspect" use this, so
that empty fields do not result in missing information when inspecting a
service.

Signed-off-by: Aaron Lehmann <aaron.lehmann@docker.com>

Commit:
Make the CLI show defaults from the swarmkit defaults package

If no fields related to an update config or restart policy are
specified, these structs should not be created as part of the service,
to avoid hardcoding the current defaults.

Signed-off-by: Aaron Lehmann <aaron.lehmann@docker.com>

vieux commented:
ping @dnephin

dnephin commented on cli/command/service/opts.go:
```diff
@@ -506,35 +586,122 @@ func addServiceFlags(flags *pflag.FlagSet, opts *serviceOptions) {
+		defaultRollbackDelay := defaults.Service.Rollback.Delay
+		rollbackDelayDesc += fmt.Sprintf(" (default %s)", defaultRollbackDelay)
+	}
+	flags.DurationVar(&opts.rollback.delay, flagRollbackDelay, 0, rollbackDelayDesc)
```
Why does the help string claim a default that isn't actually set as the default on the flag?

aaronlehmann replied:
`DurationVar` is a little weird. IIRC it shows the default value in the CLI help, but only if it's nonzero. I'd like us to always show the default on `service create`, even if it's zero, and never show it on `service update`. So it made sense to manipulate the flag description instead of providing a default value. If we did both, the default value would appear twice when the default is nonzero.

I can add a comment explaing this.

dnephin commented on cli/command/service/opts.go:
```diff
@@ -469,33 +499,83 @@ func (opts *serviceOptions) ToService() (swarm.ServiceSpec, error) {
+		service.TaskTemplate.ContainerSpec.StopGracePeriod = opts.stopGrace.Value()
+	}
+	if anyChanged(flags, flagRestartDelay, flagRestartMaxAttempts, flagRestartWindow, flagRestartCondition) {
+		service.TaskTemplate.RestartPolicy = opts.restartPolicy.defaults()
```
I would love to find a way to avoid needing `flags` here, but I can't find any right now.

Can we move this these back to the `ToRestartPolicy()` function, and add a `flags` arg ? (and the same with the few below this)

dnephin reviewed (commented):
It's unfortunate that the `pflag` library tries to stay so close to the golang `flag` package API. It is not well suited for this type of behaviour.

aaronlehmann commented on client/service_inspect.go:
```diff
@@ -3,16 +3,20 @@ package client
 // ServiceInspectWithRaw returns the service information and the raw data.
-func (cli *Client) ServiceInspectWithRaw(ctx context.Context, serviceID string) (swarm.Service, []byte, error) {
-	serverResp, err := cli.get(ctx, "/services/"+serviceID, nil, nil)
+func (cli *Client) ServiceInspectWithRaw(ctx context.Context, serviceID string, insertDefaults bool) (swarm.Service, []byte, error) {
```
SGTM. I noticed that there was a mix of styles, with some functions taking `verbose`, or `force` boolean arguments. I can switch to an options structure if that's the preferred style.

aaronlehmann commented on cli/command/service/opts.go:
```diff
@@ -190,13 +196,15 @@ type updateOptions struct {
 }
 
-func (opts updateOptions) config() *swarm.UpdateConfig {
+func (opts updateOptions) defaults(defaultUpdateConfig *api.UpdateConfig) *swarm.UpdateConfig {
```
This is passed either `defaults.Service.Update` or `defaults.Rollback.Update`. I agree that it's strange as a method on `updateOpts`. I'll change it to a function.

Any preference on whether it should be a single function that takes a `*api.UpdateConfig` (like this method), or separate functions that call a shared implementation with either the `Update` or `Rollback` default values?

dnephin replied:
Ah, I missed that, thanks!

This method moved to a function as-is sgtm.

aaronlehmann commented:
@dnephin: Thanks for the thoughtful review. I've addressed the comments - PTAL

aaronlehmann commented:
Rebased

thaJeztah commented:
I like this, finally time to give it a quick look (from a UX perspective), and seeing actual defaults in `docker service create --help` is really cool.

I haven't had time yet to test against an _actually_ older daemon, but running this;

    DOCKER_API_VERSION=1.26 docker service create --help

Seems to still show the defaults in `docker service create --help`; is that expected?

aaronlehmann commented:
> I haven't had time yet to test against an _actually_ older daemon, but running this;
> 
>     DOCKER_API_VERSION=1.26 docker service create --help
> 
> Seems to still show the defaults in `docker service create --help`; is that expected?

Yes, it's expected. I've pondered the idea of adding an API endpoint where the client can query what defaults are in place on the manager. But it seems like a lot of complexity for something that is generally not a factor (defaults change rarely, and mixed client/server version setups are unusual, and the worst case outcome is outdated help text). In this implementation, the defaults shown in CLI help are linked into the client at build time.

If you think fetching the defaults via an API is worth it, we could consider going in that direction.

aaronlehmann commented:
Rebased

dnephin reviewed (commented):
Design LGTM

aaronlehmann commented:
Rebased

dongluochen commented on docs/reference/commandline/service_update.md:
```diff
@@ -21,74 +21,73 @@ Usage:  docker service update [OPTIONS] SERVICE
       --health-interval duration           Time between running the check (ns|us|ms|s|m|h)
       --health-retries int                 Consecutive failures needed to report unhealthy
       --health-timeout duration            Maximum time to allow one check to run (ns|us|ms|s|m|h)
       --health-start-period duration       Start period for the container to initialize before counting retries towards unstable (ns|us|ms|s|m|h) (default 0s)
```
remove `(default 0s)`

aaronlehmann replied:
Fixed

dongluochen reviewed (commented):
sgtm

dongluochen commented:
design lgtm.

vdemeester reviewed (commented):
Design LGTM, moving to code review

vdemeester commented:
@aaronlehmann needs a rebase 👼

aaronlehmann commented:
Rebased. `flagUpdateOrder` and `flagRollbackOrder` are now handled.

vieux commented:
@aaronlehmann seems like the windows failure is related
`..\..\api\server\router\swarm\helpers.go:42: not enough arguments in call to sr.backend.GetService`

aaronlehmann commented:
@vieux: Looks like the Windows build was broken by #32154. It's not related to this change.

aaronlehmann commented:
Deleted my earlier comment, because it was indeed a problem with this PR. I thought it was an existing problem on master, but it turned out the reason it only showed up on windows was because each builder does a separate merge to master so they're not building the same version of the code (which is kind of a problem IMHO!).

aaronlehmann commented:
@vieux: Fixed, thanks!

vdemeester reviewed (approved):
LGTM 🐯

thaJeztah commented:
I'm a bit confused, but probably do something wrong..

```bash
docker service create --restart-condition="on-failure" --name web nginx:alpine
2i3pres421je9kuo5rh1vf2n4
```

Makes this request;

```json
{
  "EndpointSpec": {
    "Mode": "vip"
  },
  "Labels": {},
  "Mode": {
    "Replicated": {}
  },
  "Name": "web",
  "TaskTemplate": {
    "ContainerSpec": {
      "DNSConfig": {},
      "Image": "nginx:alpine"
    },
    "ForceUpdate": 0,
    "Placement": {},
    "Resources": {
      "Limits": {},
      "Reservations": {}
    },
    "RestartPolicy": {
      "Condition": "on-failure",
      "Delay": 5000000000,
      "MaxAttempts": 0
    }
  }
}
````


With the `insertDefaults=true` option set;

```bash
/ # curl --unix-socket /var/run/docker.sock -sL http://localhost/v1.29/services/web?insertDefaults=true | jq .
{
  "ID": "2i3pres421je9kuo5rh1vf2n4",
  "Version": {
    "Index": 31
  },
  "CreatedAt": "2017-04-10T09:36:58.859857088Z",
  "UpdatedAt": "2017-04-10T09:36:58.859857088Z",
  "Spec": {
    "Name": "web",
    "Labels": {},
    "TaskTemplate": {
      "ContainerSpec": {
        "Image": "nginx:alpine@sha256:5aadb68304a38a8e2719605e4e180413f390cd6647602bee9bdedd59753c3590",
        "DNSConfig": {}
      },
      "Resources": {
        "Limits": {},
        "Reservations": {}
      },
      "RestartPolicy": {
        "Condition": "on-failure",
        "Delay": 5000000000,
        "MaxAttempts": 0
      },
      "Placement": {},
      "ForceUpdate": 0
    },
    "Mode": {
      "Replicated": {
        "Replicas": 1
      }
    },
    "EndpointSpec": {
      "Mode": "vip"
    }
  },
  "Endpoint": {
    "Spec": {}
  }
}
```

Without the `insertDefaults=true` option set;

```bash
/ # curl --unix-socket /var/run/docker.sock -sL http://localhost/v1.29/services/web | jq .
{
  "ID": "2i3pres421je9kuo5rh1vf2n4",
  "Version": {
    "Index": 31
  },
  "CreatedAt": "2017-04-10T09:36:58.859857088Z",
  "UpdatedAt": "2017-04-10T09:36:58.859857088Z",
  "Spec": {
    "Name": "web",
    "Labels": {},
    "TaskTemplate": {
      "ContainerSpec": {
        "Image": "nginx:alpine@sha256:5aadb68304a38a8e2719605e4e180413f390cd6647602bee9bdedd59753c3590",
        "DNSConfig": {}
      },
      "Resources": {
        "Limits": {},
        "Reservations": {}
      },
      "RestartPolicy": {
        "Condition": "on-failure",
        "Delay": 5000000000,
        "MaxAttempts": 0
      },
      "Placement": {},
      "ForceUpdate": 0
    },
    "Mode": {
      "Replicated": {
        "Replicas": 1
      }
    },
    "EndpointSpec": {
      "Mode": "vip"
    }
  },
  "Endpoint": {
    "Spec": {}
  }
}
```

aaronlehmann commented:
@thaJeztah: You're right - there was a problem. Specifying the service ID led to the correct behavior, but services names were not handled properly, as the service was returned directly from `ListServices`. I've fixed this and added test coverage.

thaJeztah commented:
Ah! That looks better;

```bash
docker service inspect web
[
    {
        "ID": "k850seipz5ohpemq1cotenze2",
        "Version": {
            "Index": 10
        },
        "CreatedAt": "2017-04-11T10:51:21.181483381Z",
        "UpdatedAt": "2017-04-11T10:51:21.181483381Z",
        "Spec": {
            "Name": "web",
            "Labels": {},
            "TaskTemplate": {
                "ContainerSpec": {
                    "Image": "nginx:alpine@sha256:5aadb68304a38a8e2719605e4e180413f390cd6647602bee9bdedd59753c3590",
                    "StopGracePeriod": 10000000000,
                    "DNSConfig": {}
                },
                "Resources": {
                    "Limits": {},
                    "Reservations": {}
                },
                "RestartPolicy": {
                    "Condition": "on-failure",
                    "Delay": 5000000000,
                    "MaxAttempts": 0
                },
                "Placement": {},
                "ForceUpdate": 0
            },
            "Mode": {
                "Replicated": {
                    "Replicas": 1
                }
            },
            "UpdateConfig": {
                "Parallelism": 1,
                "FailureAction": "pause",
                "Monitor": 5000000000,
                "MaxFailureRatio": 0,
                "Order": "stop-first"
            },
            "RollbackConfig": {
                "Parallelism": 1,
                "FailureAction": "pause",
                "Monitor": 5000000000,
                "MaxFailureRatio": 0,
                "Order": "stop-first"
            },
            "EndpointSpec": {
                "Mode": "vip"
            }
        },
        "Endpoint": {
            "Spec": {}
        }
    }
]
```

thaJeztah commented on api/swagger.yaml:
```diff
@@ -7584,6 +7584,11 @@ paths:
           description: "ID or name of service."
           required: true
           type: "string"
+        - name: "insertDefaults"
```
This needs to be added to the API version-history; https://github.com/docker/docker/blob/master/docs/api/version-history.md

(ok with doing that in a follow-up)

thaJeztah reviewed (approved):
LGTM

thaJeztah commented:
All green; I'll go ahead and merge so that this is in 17.05. I'll include the API version changes https://github.com/docker/docker/pull/32284#discussion_r110872635 in the "changelog" pull request
